numpy/core/src/multiarray/methods.c

```diff
@@ -856,20 +856,30 @@ array_astype(PyArrayObject *self,
      * and it's not a subtype if subok is False, then we
      * can skip the copy.
      */
-    if (!forcecopy && (order == NPY_KEEPORDER ||
-                       (order == NPY_ANYORDER &&
-                            (PyArray_IS_C_CONTIGUOUS(self) ||
-                            PyArray_IS_F_CONTIGUOUS(self))) ||
-                       (order == NPY_CORDER &&
-                            PyArray_IS_C_CONTIGUOUS(self)) ||
-                       (order == NPY_FORTRANORDER &&
-                            PyArray_IS_F_CONTIGUOUS(self))) &&
-                    (subok || PyArray_CheckExact(self)) &&
-                    PyArray_EquivTypes(dtype, PyArray_DESCR(self))) {
+    if (forcecopy != NPY_COPY_ALWAYS && 
+        (order == NPY_KEEPORDER ||
+        (order == NPY_ANYORDER &&
+            (PyArray_IS_C_CONTIGUOUS(self) ||
+            PyArray_IS_F_CONTIGUOUS(self))) ||
+        (order == NPY_CORDER &&
+            PyArray_IS_C_CONTIGUOUS(self)) ||
+        (order == NPY_FORTRANORDER &&
+            PyArray_IS_F_CONTIGUOUS(self))) &&
+    (subok || PyArray_CheckExact(self)) &&
+    PyArray_EquivTypes(dtype, PyArray_DESCR(self))) {
         Py_DECREF(dtype);
         Py_INCREF(self);
         return (PyObject *)self;
     }
+
+    if( forcecopy == NPY_COPY_NEVER ) {
+        PyErr_SetString(PyExc_ValueError,
+                        "Unable to avoid copy while casting in "
+                        "np._CopyMode.NEVER");
```

Comment:
Again, don't mention a private variable in the error message.  Otherwise, I guess this is probably right (although we may have to modify it at some point, it would be nice to allow view-casts, but that is a different story).